pkg/machines/libvirt-dbus.es6

```diff
@@ -0,0 +1,861 @@
+    vmActionFailed,
+} from './actions/store-actions.es6';
+
+import {
+    getDiskXML,
+    getVolumeXML
+} from './xmlCreator.es6';
+
+import {
+    usagePollingEnabled
+} from './selectors.es6';
+
+import VCPUModal from './components/vcpuModal.jsx';
+
+import {
+    logDebug
+} from './helpers.es6';
+
+import {
+    canConsole,
+    canDelete,
+    canInstall,
+    canReset,
+    canRun,
+    canSendNMI,
+    canShutdown,
+    getSingleOptionalElem,
+    isRunning,
+    parseDumpxml,
+    serialConsoleCommand,
+    unknownConnectionName,
+    updateVCPUSettings,
+    CONSOLE_VM,
+    CHECK_LIBVIRT_STATUS,
+    CREATE_VM,
+    ENABLE_LIBVIRT,
+    GET_OS_INFO_LIST,
+    INIT_DATA_RETRIEVAL,
+    INSTALL_VM,
+    START_LIBVIRT,
+} from './libvirt-common.es6';
+
+const _ = cockpit.gettext;
+
+let clientLibvirt = {};
+/* Default timeout for libvirt-dbus method calls */
+const TIMEOUT = { timeout: 30000 };
+
+let LIBVIRT_DBUS_PROVIDER = {};
+LIBVIRT_DBUS_PROVIDER = {
+    name: 'LibvirtDBus',
+
+    /*
+     * Initialize the provider.
+     *
+     * @param providerContext
+     * @returns {boolean} - true, if initialization succeeded; or Promise
+     */
+    init(providerContext) {
+        return true;
+    },
+
+    openVCPUModal: (params) => VCPUModal(params),
+
+    /* Start of common provider functions */
+    canConsole,
+    canDelete,
+    canInstall,
+    canReset,
+    canRun,
+    canSendNMI,
+    canShutdown,
+    isRunning,
+    serialConsoleCommand,
+    CONSOLE_VM,
+    CHECK_LIBVIRT_STATUS,
+    CREATE_VM,
+    ENABLE_LIBVIRT,
+    GET_OS_INFO_LIST,
+    INIT_DATA_RETRIEVAL,
+    INSTALL_VM,
+    START_LIBVIRT,
+    /* End of common provider functions  */
+
+    ATTACH_DISK({
+        connectionName,
+        diskFileName,
+        target,
+        vmId,
+        permanent,
+        hotplug
+    }) {
+        /*
+         * 0 -> VIR_DOMAIN_AFFECT_CURRENT
+         * 1 -> VIR_DOMAIN_AFFECT_LIVE
+         * 2 -> VIR_DOMAIN_AFFECT_CONFIG
+         */
+        let flags = 0;
+        if (hotplug)
+            flags |= 1;
+        if (permanent)
+            flags |= 2;
+
+        let xmlDesc = getDiskXML(diskFileName, target);
+
+        return () => {
+            clientLibvirt[connectionName].call(vmId, 'org.libvirt.Domain', 'AttachDevice', [xmlDesc, flags], TIMEOUT)
+                    .fail(ex => console.error("VM ATTACH DISK action for disk %s failed: %s", diskFileName, JSON.stringify(ex)));
+        };
+    },
+
+    CHANGE_NETWORK_STATE({
+        name,
+        id: objPath,
+        networkMac,
+        state,
+        connectionName
+    }) {
+        return dispatch => {
+            clientLibvirt[connectionName].call(objPath, 'org.libvirt.Domain', 'GetXMLDesc', [0], TIMEOUT)
+                    .done(domXml => {
+                        let updatedXml = updateNetworkIfaceState(domXml[0], networkMac, state);
+                        if (!updatedXml) {
+                            dispatch(vmActionFailed({
```

Comment:
.. so I suppose that's one half of the failure handler, if the dbus call succeeds, but returns an empty XML. But it could also fail, no?

Reply:
Fixed